### DISCUSSION TITLE
Maven not existing for Github Actions?

### DISCUSSION THREAD

**Discussion by UnfamousThomas** [community/community#25734] [Actions]
<p>Hello.</p>
<p>I am writing regarding an issue I am having on a self-hosted runner in Github Actions. I want to see if I am missing something, or am just being stupid - hopefully someone here knows what my issue is.</p>
<p>Let’s get started.</p>
<p>Everything was fine before I decided to update my Java project (using Maven) from java 8 to java 17. After which, a bunch of issues started. I eventually tracked down that the issue came from the fact that the Maven I got from the <code>apt</code> package manager was an older version (3.6.x) while for java 17, I needed a recent version (3.8.x). Now… The apt package manager does not have that, so I had to go through the manual installation process for it.</p>
<p>I used <a href="https://computingforgeeks.com/install-latest-apache-maven-on-ubuntu-debian/" rel="noopener nofollow ugc">this page here</a> as a guide/tutorial and got it working from shell, at least.</p>
<p>That’s where the main issue comes in. Github Actions cannot seem to access the ‘mvn’ command (throws error regarding mvn cmd not being found). However, when I use either root or the user which runner is under manually from ssh, it does actually exist and work.</p>
<p>I am using Ubuntu 18.04.</p>
<p>Here is my <code>mvn -v</code> output:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/5/7/57108184776e54c868c93c684f773ce4d210b9a1.png" data-download-href="/uploads/short-url/cqcU1SKRpl4VTyMeEAyKxky0mid.png?dl=1" title="mvnver" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/31453763/181092941-891d6a3b-c0bf-4690-a2f1-75c9a63dd4bc.png" alt="mvnver" data-base62-sha1="cqcU1SKRpl4VTyMeEAyKxky0mid" width="690" height="78" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/5/7/57108184776e54c868c93c684f773ce4d210b9a1_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">mvnver</span><span class="informations">752×86 4.79 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>And Github Actions error:</p>
<pre><code class="lang-auto">Run mvn clean install deploy -B "-DcustomVersion=dev-33"
/home/operations/actions-runner/_work/_temp/a9f0474a-e306-44d5-8093-0473929c5cb8.sh: line 1: mvn: command not found
Error: Process completed with exit code 127.
</code></pre>
<p>Finally, my workflow file. I don’t know how exactly the text formatting work here, so apologies if the spacing gets messed up.</p>
<pre><code class="lang-auto">on:
  push:
    branches:
      - dev
      - release
      - alpha
  workflow_dispatch:

jobs:
  build:
    runs-on: self-hosted

    steps:
      - uses: actions/checkout@v1

      - name: Get branch name
        id: branch-name
        uses: tj-actions/branch-names@v5.1

      - name: Cache Maven packages
        id: cache
        uses: actions/cache@v2
        with:
          path: ~/.m2
          key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
          restore-keys: ${{ runner.os }}-m2
          
      - name: Setup settings.xml
        uses: s4u/maven-settings-action@v2.4.1
        with:
          servers: |
            [
               {
                  "id":"x3",
                  "username":"${{ secrets.ARTIFACTS_USER }}",
                  "password":"${{ secrets.ARTIFACTS_PASS }}"
               },
               {
                  "id":"x2",
                  "username":"${{ secrets.ARTIFACTS_USER }}",
                  "password":"${{ secrets.ARTIFACTS_PASS }}"
               },
               {
                  "id":"x1",
                  "username":"${{ secrets.ARTIFACTS_USER }}",
                  "password":"${{ secrets.ARTIFACTS_PASS }}"
               }
            ]
      - name: Build with Maven
        run: mvn clean install deploy -B "-DcustomVersion=${{ steps.branch-name.outputs.current_branch }}-${{ github.run_number }}"
</code></pre>

**Comment by ViacheslavKudinov** [comment, marked as answer]
<p>Hi <a class="mention" href="/unfamousthomas">@UnfamousThomas</a><br>
according to your workflow, seems you run actions on self-hosted runners.<br>
Could it be that mvn is not in the PATH of user from which run server (runner application)?<br>
Could you try to add a full path , eg <code>/some_path/mvn</code> in your workflow instead of just <code>mvn</code>?</p>

**Comment by UnfamousThomas** [comment]
<p>Hey.</p>
<p>Just tested it, this did indeed work… Though it is a bit strange that it did to me.</p>
<p>When I run <code>mvn</code> or <code>echo $PATH</code> in the <code>runner</code> user, they both display the correct outputs… Well, still. Thanks. This solves my issues, having some extra paths at the cmd is not really an issue.</p>

**Comment by ViacheslavKudinov** [comment]
<p>Probably, you can also try to set PATH variable once in the workflow.<br>
Something like:</p>
<pre><code class="lang-auto">steps:
  - name: Set path to maven
    id: mvn_path
    run: |
      echo "/opt/maven/bin" &gt;&gt; $GITHUB_ENV
</code></pre>
<p>more info <a href="https://docs.github.com/en/actions/learn-github-actions/workflow-commands-for-github-actions#adding-a-system-path" class="inline-onebox" rel="noopener nofollow ugc">Workflow commands for GitHub Actions - GitHub Docs</a><br>
If it works fine you can have short <code>mvn</code> in all other steps you need to execute it.</p>

**Comment by ViacheslavKudinov** [comment]
<p><a class="mention" href="/unfamousthomas">@UnfamousThomas</a><br>
If i remember correctly i’ve added symlink to make it visible to runner app.<br>
<code>sudo ln -s /opt/apache-maven-${MAVEN_VERSION}/bin/mvn /usr/bin/mvn</code><br>
You can change PATH to your, where is located mvn.</p>

**Comment by ViacheslavKudinov** [comment]
<p>Also could be this is better than support own setup of Maven - <a href="https://github.com/marketplace/actions/setup-maven" class="inline-onebox" rel="noopener nofollow ugc">Setup Maven · Actions · GitHub Marketplace · GitHub</a><br>
This lines of code <a href="https://github.com/stCarolas/setup-maven/blob/7060d72b50fbf2389a16d6f66a9c7d4d9ece5d36/src/installer.ts#L30L31" class="inline-onebox" rel="noopener nofollow ugc">setup-maven/installer.ts at 7060d72b50fbf2389a16d6f66a9c7d4d9ece5d36 · stCarolas/setup-maven · GitHub</a> make it available for the runner app.</p>

**Comment by UnfamousThomas** [comment]
<p>This should indeed work in theory, but for me at least, does not.</p>
<p>As for the other options (PATH variable in workflow) - I have not really tested it, I originally opted to supply the mvn path via a secret instead. And then do <code>${{ secrets.MVN_PATH }}/mvn subcommand</code>.</p>
<p>Finally, the action (setup maven) option. Yeah, that works. Thanks… I didn’t really even consider that option.</p>

**Comment by jsoref** [comment]
<p>I’d suggest <em>not</em> using a secret for non secret paths.</p>
<p>GitHub will redact secrets in logs, and that will make it harder for you to read them.</p>
<p>An action, another repository, … anything else would be better.</p>
